### Brief Overview 

- The goal of this assignment is to practice some important data wrangling functionality commonly required in real-world projects.

- Here we will use two datasets:
  - IRS Statistics of Income (SOI) dataset
  - The Medicaid Data per State  


- The final product here is a table with medication cost per Medicaid enrollee per state. This dataset will allow us to answer such questions as:
  - Which medications account for the bulk of a state's spending   
  - Which drugs are prescribed much more in one state compared to the other states.
etc.



### Instructions on answering the questions

- Some of the questions below require that you only use methods or properties of a `DataFrame` or a `Series`. Therefore, any solution that uses a function that is not a method or property of `DataFrame`s or `Series` will not be accepted, even when the solution yields an appropriate answer to the question.

- For instance, if you are asked to find the number of entries in the DataFrame `tax_data ` using only the DataFrame's methods or properties, then `len(tax_data)` is not an acceptable solution since `len()` is not a `tax_data` method. The statements below are both correct answers:

  - `info()` is a method `tax_data`
  
```python
   
    tax_data.info()
```

  or

  - `shape` is a property of `tax_data`


```python
tax_data.shape
```

- Similarly, if you are asked to find to count the number of unique entries in the `STATEFIPS` column of the `tax_data` DataFrame, then solutions using set() or len() are not acceptable for the following reasons:

- The solution below uses `set()` and `len()`, which are not `tax_data` methods

```python
len(set(tax_data["STATEFIPS"]))
```

- The solution below uses `unique()`, which is a  `tax_data` method, but then counts the number of uniques entries using `len()` which is not `tax_data` method

```python
len(tax_data["STATEFIPS"].unique())
```


- The statement below is an acceptable solution since it uses `nunique()` which is a method of the `Series` generated by indexing on a column of `tax_data` (`tax_data['STATEFIPS'])

```python
tax_data['STATEFIPS'].nunique()
```


- Chaining methods and properties is encouraged if it does not cause ambiguity 

- For instance, to identify whether a value is part of the index, write the following:

```python
tax_data.index.contains(99999)
```

- Rather than:

```python
9999 in tax_data.index
```

- You can only import `pandas` and `numpy`

- If you are not explicitely asked to only use methods or properties of a `DataFrame` or a `Series`, then any solution that does not rely on external products will be accepted.

- __Important__: Provide the exact statement(s) used to answer each question

- Unless otherwise specified, each cell can contain multiple lines of code

* Finally, note that not all the functions necessary for answering the questions below were covered in class. As such, I suggest you use `.SHIFT+TAB` on objects liberally to see which methods and properties are available on objects. If you are unsure what a method does, use `SHIFT+TAB` twice to invoke the `docstring`, or documentation for that function. This is not only a good way to see which functionality can be used to answer the questions below but also a great way to familiarize yourself with the plethora or functionality available through the `pandas` package.

---

In [ ]:
import pandas as pd
import numpy as np


* Load the IRS Statistics of Income (SOI) dataset (tax_data.csv) into a `DataFrame` called `tax_data`. The file is `tax_data.csv` is located in the `data` directory of the assignment folder.

* This dataset was preprocessed but the original one was obtained at the following URL:

        https://www.irs.gov/pub/irs-soi/15zpallagi.csv


In [2]:
tax_data  = pd.read_csv("data/tax_data.csv")
tax_data

STATEFIPS STATE  zipcode  agi_stub        N1     mars1     MARS2  \
0             1.0    AL      0.0       1.0  836320.0  481570.0  109790.0   
1             1.0    AL      0.0       2.0  494830.0  206630.0  146250.0   
2             1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   
3             1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   
4             1.0    AL      0.0       5.0  212660.0   19520.0  184320.0   
5             1.0    AL      0.0       6.0   55360.0    2950.0   49260.0   
6             1.0    AL  35004.0       1.0    1490.0     970.0     230.0   
7             1.0    AL  35004.0       2.0    1350.0     630.0     360.0   
8             1.0    AL  35004.0       3.0     970.0     310.0     490.0   
9             1.0    AL  35004.0       4.0     620.0     110.0     470.0   
10            1.0    AL  35004.0       5.0     620.0      40.0     560.0   
11            1.0    AL  35004.0       6.0      60.0       0.0      50.0   
12            1.0    AL  35005.0       1.0    1350.0     750.0     190.0   
13            1.0    AL  35005.0       2.0     980.0     370.0     230.0   
14            1.0    AL  35005.0       3.0     490.0     150.0     210.0   
15            1.0    AL  35005.0       4.0     250.0      50.0     180.0   
16            1.0    AL  35005.0       5.0     190.0       0.0     160.0   
17            1.0    AL  35005.0       6.0       0.0       0.0       0.0   
18            1.0    AL  35006.0       1.0     430.0     260.0     100.0   
19            1.0    AL  35006.0       2.0     340.0     120.0     150.0   
20            1.0    AL  35006.0       3.0     190.0      30.0     130.0   
21            1.0    AL  35006.0       4.0     130.0      30.0     100.0   
22            1.0    AL  35006.0       5.0     140.0       0.0     130.0   
23            1.0    AL  35006.0       6.0       0.0       0.0       0.0   
24            1.0    AL  35007.0       1.0    4170.0    2750.0     560.0   
25            1.0    AL  35007.0       2.0    2700.0    1160.0     840.0   
26            1.0    AL  35007.0       3.0    1840.0     570.0     950.0   
27            1.0    AL  35007.0       4.0    1440.0     190.0    1140.0   
28            1.0    AL  35007.0       5.0    1780.0     130.0    1590.0   
29            1.0    AL  35007.0       6.0     240.0       0.0     220.0   
...           ...   ...      ...       ...       ...       ...       ...   
166668       56.0    WY  83126.0       1.0      60.0      60.0      40.0   
166669       56.0    WY  83126.0       2.0      30.0       0.0       0.0   
166670       56.0    WY  83126.0       3.0      40.0       0.0      30.0   
166671       56.0    WY  83126.0       4.0      40.0       0.0      40.0   
166672       56.0    WY  83126.0       5.0       0.0       0.0       0.0   
166673       56.0    WY  83126.0       6.0       0.0       0.0       0.0   
166674       56.0    WY  83127.0       1.0     380.0     240.0      80.0   
166675       56.0    WY  83127.0       2.0     300.0     120.0     150.0   
166676       56.0    WY  83127.0       3.0     230.0      70.0     150.0   
166677       56.0    WY  83127.0       4.0     190.0      30.0     160.0   
166678       56.0    WY  83127.0       5.0     230.0      20.0     200.0   
166679       56.0    WY  83127.0       6.0      70.0       0.0      60.0   
166680       56.0    WY  83128.0       1.0     250.0     190.0      30.0   
166681       56.0    WY  83128.0       2.0     210.0     120.0      60.0   
166682       56.0    WY  83128.0       3.0     160.0      70.0      90.0   
166683       56.0    WY  83128.0       4.0      90.0       0.0      80.0   
166684       56.0    WY  83128.0       5.0     100.0      30.0      90.0   
166685       56.0    WY  83128.0       6.0      50.0       0.0      40.0   
166686       56.0    WY  83414.0       1.0      60.0      50.0       0.0   
166687       56.0    WY  83414.0       2.0      60.0      40.0      50.0   
166688       56.0    WY  83414.0       3.0       0.0    

* Use a `tax_data` method or property to display the first eight (8) rows of the `DataFrame`

In [3]:
tax_data.head(8)

STATEFIPS STATE  zipcode  agi_stub        N1     mars1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   
5        1.0    AL      0.0       6.0   55360.0    2950.0   49260.0     350.0   
6        1.0    AL  35004.0       1.0    1490.0     970.0     230.0     280.0   
7        1.0    AL  35004.0       2.0    1350.0     630.0     360.0     300.0   

       PREP         N2    ...        N10300     A10300   N85530   A85530  \
0  455560.0  1356760.0    ...      373410.0   328469.0      0.0      0.0   
1  275920.0  1010990.0    ...      395880.0   965011.0      0.0      0.0   
2  155100.0   583910.0    ...      251490.0  1333418.0      0.0      0.0   
3   99950.0   423990.0    ...      165320.0  1414283.0      0.0      0.0   
4  126860.0   589490.0    ...      212000.0  3820152.0    420.0    168.0   
5   41410.0   160530.0    ...       55300.0  6027793.0  22090.0  39519.0   
6     700.0     2160.0    ...         690.0      610.0      0.0      0.0   
7     610.0     2540.0    ...        1140.0     3019.0      0.0      0.0   

    N85300   A85300   N11901    A11901    N11902     A11902  
0      0.0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0      0.0  45460.0  128823.0  116440.0   377177.0  
4     60.0     31.0  83330.0  421004.0  121570.0   483682.0  
5  27550.0  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0      0.0    120.0      94.0    1290.0     2792.0  
7      0.0      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 91 columns]

*  Modify `tax_data` to uppercase all the header name. 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself
  *  You can `tax_data.columns`, which returns a `Series` of the column names.
  
* The resulting column name should look as follows:


```
STATEFIPS    STATE    ZIPCODE    AGI_STUB    N1    MARS1    MARS2    MARS4    PREP    N2    ...    
```

This operation is useful for standardizing column names and avoid guessing whether the column header was in upper case, lower case or a mix of both.

In [4]:
tax_data.columns = map(str.upper, tax_data.columns)
tax_data.head(8)

STATEFIPS STATE  ZIPCODE  AGI_STUB        N1     MARS1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   
5        1.0    AL      0.0       6.0   55360.0    2950.0   49260.0     350.0   
6        1.0    AL  35004.0       1.0    1490.0     970.0     230.0     280.0   
7        1.0    AL  35004.0       2.0    1350.0     630.0     360.0     300.0   

       PREP         N2    ...        N10300     A10300   N85530   A85530  \
0  455560.0  1356760.0    ...      373410.0   328469.0      0.0      0.0   
1  275920.0  1010990.0    ...      395880.0   965011.0      0.0      0.0   
2  155100.0   583910.0    ...      251490.0  1333418.0      0.0      0.0   
3   99950.0   423990.0    ...      165320.0  1414283.0      0.0      0.0   
4  126860.0   589490.0    ...      212000.0  3820152.0    420.0    168.0   
5   41410.0   160530.0    ...       55300.0  6027793.0  22090.0  39519.0   
6     700.0     2160.0    ...         690.0      610.0      0.0      0.0   
7     610.0     2540.0    ...        1140.0     3019.0      0.0      0.0   

    N85300   A85300   N11901    A11901    N11902     A11902  
0      0.0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0      0.0  45460.0  128823.0  116440.0   377177.0  
4     60.0     31.0  83330.0  421004.0  121570.0   483682.0  
5  27550.0  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0      0.0    120.0      94.0    1290.0     2792.0  
7      0.0      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 91 columns]

* What is the total number of entries (also called observations) in `tax_data`?

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [5]:
tax_data.size

15169518

- If `STATEFIPS` is header title of the first column of the `tax_data` `DataFrame`, what is the title of the 32nd column
  - Your answer can only use `DataFrame` or `Series` methods or properties and should use a single python expression



In [6]:
tax_data.columns.values[31]

'N00900'

* If `STATEFIPS` is the the first column, what is the index of the column name `N10300`?

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [7]:
tax_data.columns.get_loc("N10300")

81

- Print the count of unique zip codes in each state using descending order.
  - Your answer can only use `DataFrame` or `Series` methods or properties

- The result should look like the following ( '...' represents remaining data that is not shown)
```python

        ZIPCODE
STATE	
TX	     9714
NY	     9238
CA	     8908
PA	     8208
IL	     7386
...
```

* The above indicates that there are 9,714 unique zipcodes in TX, 9,238 in NY, etc.


In [8]:
#unique_zip = tax_data.groupby('STATE').count().loc[:, ['ZIPCODE']].sort_values(by=['ZIPCODE'], ascending=False)
#unique_zip
tax_data.groupby('STATE')['ZIPCODE'].nunique().sort_values(ascending=False)


STATE
TX    1619
NY    1540
CA    1485
PA    1368
IL    1231
OH     998
FL     920
MI     892
MO     890
IA     827
VA     793
MN     792
NC     725
WI     714
IN     674
GA     667
KY     656
KS     598
TN     590
AL     577
OK     547
NJ     547
WV     512
WA     497
AR     490
NE     486
MA     482
LA     453
MD     405
CO     395
SC     376
ME     370
MS     370
OR     356
AZ     295
ND     287
SD     287
CT     264
VT     238
NH     232
MT     225
ID     213
NM     209
UT     186
NV     128
WY     109
RI      71
HI      60
DE      57
AK      56
DC      24
Name: ZIPCODE, dtype: int64

- Identify the position of HI in the list of zip code counts per state (questions directly above)
  - Your answer can only use `DataFrame` or `Series` methods or properties

In [9]:
unique_zip.index.get_loc('HI')


NameError: name 'unique_zip' is not defined

### Identifying and Removing Ambiguous Zip Codes

- Count the number of entries where ZIPCODE is 0, assign your results to a variable named  `nb_invalid_zip`

In [ ]:
nb_invalid_zip = tax_data['ZIPCODE'].eq(0).sum().astype(int)
nb_invalid_zip.astype(int)
nb_invalid_zip
print(nb_invalid_zip)
print(type(nb_invalid_zip))


* Run the line below to make sure that `nb_invalid_zip` is an integer (`int`)
  * Note that `assert` will only print an error if `type(nb_invalid_zip)` is not of type `int`
  
* If `nb_invalid_zip` then change your answer above so that the value returned is effectively a number.

In [ ]:
assert(type(nb_invalid_zip) == int)

- Remove from `tax_data` all the lines where the zip code is `0` and save resulting `DataFrame` to a variable named `tax_data_valid_zip`
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [ ]:
tax_data_valid_zip = tax_data[tax_data.ZIPCODE != 0]
tax_data_valid_zip

* Run the line below to confirm that the operation worked as expected

  * The assertion below is testing that the number of lines with "zip code equal to  0" + number of lines in `tax_data_valid_zip` is equal to the number of lines in the original `DataFrame` `tax_data`
  
  * The assertion below will fail (and print an error message) if the results do not match. If that is the case, please review your code above.

In [ ]:
assert((tax_data_valid_zip.shape[0] + nb_invalid_zip) == tax_data.shape[0])

### Identifying and Removing Lines with Missing Values

* How many lines contain at least one missing value `NaN` in the `tax_data_valid_zip` `DataFrame`?
  * Your answer can only use `DataFrame` methods and properties
* Assing the count of `NaN` into a variable called nb_missing_values

In [ ]:
nb_missing_values = tax_data_valid_zip.loc[:, :].isnull().sum()

nb_missing_values = nb_missing_values.loc[:].sum()

nb_missing_values

* Create a new `DataFrame` containing all the lines from `tax_data_valid_zip`, except lines containing missing values

* Call the new `DataFrame` `tax_data_valid_zip_cleaned

In [ ]:
tax_data_valid_zip_cleaned = tax_data_valid_zip.dropna()
tax_data_valid_zip_cleaned

* Run the line below to confirm that the operation worked as expected. The assertion below is testing that:  
`nb_missing_values` + number of lines in `tax_data_valid_zip_cleaned` is equal to the number of lines in `tax_data_valid_zip`
  
* Note that assert will only print an error if the results do not match

In [ ]:
assert((tax_data_valid_zip_cleaned.shape[0] + nb_missing_values) == tax_data_valid_zip.shape[0])

### Computing the Percentile Income per Zip Code

* The function `compute_percentile_zipcode` below computes the percentile income per zip code

* By default percentile=0.5,  i.e., the function computes the median

* Read the code and make sure you understand what it does before moving on to the next question


In [ ]:
def compute_percentile_zip(df_zip, percentile=0.65):     
    index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

* Compute the 65th income percentile ( percentile=0.65) for each zipcode in `tax_data_valid_zip_cleaned` data frame

  * Your answer can only use `DataFrame` or `Series` methods or properties
  
  * Recall that you can use the `apply` function to transform the values of a column
  
  
* Sort the values in descending order and assign them to a new `DataFrame` called `zip_rev_all`
* The resulting `Series` should look like the following( '...' represents remaining data that is not shown)

```Python
ZIPCODE
33109    3954.114286
33480    3413.301538
...
```

In [ ]:
zip_rev_all = tax_data_valid_zip_cleaned.groupby('ZIPCODE').apply(compute_percentile_zip).sort_values(ascending=False)
zip_rev_all

- What are the three zip codes with the most significant 65th percentile value for income?

In [ ]:
zip_rev_all.head(3)

# 2 Working with the Medicaid Data

### Loading and exploring the data 

* Load the Medicaid data stored in the file `medicaid_data.csv` into a `DataFrame` called `medicaid_data`. The file is located in the `data` directory of the assignment folder. 
* Note that this is quite large and may take some time to load on a computer with modest RAM resources (4GB or less)


In [10]:
medicaid_data  = pd.read_csv("data/medicaid_data.csv")
medicaid_data

Utilization Type State          NDC Product Name  Units Reimbursed  \
0                   MCOU    PA  55150023930   Dexamethas            33.000   
1                   FFSU    NY     23917710   ALPHAGAN P           570.000   
2                   MCOU    OR  13925050501   Dapsone 10           456.000   
3                   FFSU    MN  51862006401     DIAZEPAM           780.000   
4                   FFSU    MN    781237101   DEXTROAMPH           451.000   
5                   FFSU    VT    603786274   TRIAMCINOL           315.000   
6                   FFSU    NY  24385060271   CHEST CONG          4773.000   
7                   FFSU    MO    603024118       Q-DRYL          9581.000   
8                   FFSU    CT  62756026002   ZONISAMIDE          2530.000   
9                   FFSU    OK  55111065301    Sirolimus           750.000   
10                  FFSU    KY    406012401   HYDROCODON          3564.000   
11                  MCOU    GA  68382000414   Topiramate          8218.000   
12                  FFSU    MA    406012501   HYDROCODON         15732.000   
13                  FFSU    XX    904657561   PHENOBARBI          1576.000   
14                  FFSU    LA  43199003701    LABETALOL          1064.000   
15                  FFSU    MO    472524267   PERMETHRIN          2068.000   
16                  FFSU    ID  67877014601    TEMAZEPAM          2990.000   
17                  FFSU    NV    173087310   INCRUSE EL          1050.000   
18                  MCOU    NY  68645051070   HYDROCHLOR         71982.000   
19                  FFSU    MD    378306677   FENOFIBRAT           724.000   
20                  FFSU    PA    456060010   TEFLARO 60           224.000   
21                  FFSU    XX  18860010501   FAZACLO 15          4206.000   
22                  FFSU    OH  69097015815    MELOXICAM          3426.000   
23                  FFSU    MT    185007460   LOVASTATIN          1421.000   
24                  MCOU    OH    409488799   STERILE WA          4054.601   
25                  MCOU    VA  59762100501   Cabergolin           104.000   
26                  MCOU    FL  29300011316   LAMOTRIGIN           872.000   
27                  FFSU    TN      3633717       DROXIA          6187.000   
28                  FFSU    IN     69046956   Chantix (v           890.000   
29                  FFSU    SC    603497521   OXYBUTYNIN          2747.000   
...                  ...   ...          ...          ...               ...   
1695516             MCOU    MI    168000615   TRIAMCINOL         71290.000   
1695517             MCOU    MI  16714069501   FLUCONAZOL          7340.000   
1695518             MCOU    NE    173069500   ADVAIR DIS         26520.000   
1695519             FFSU    OK  59762444002   Methylpred         17152.000   
1695520             MCOU    FL     54002511   MEFLOQUINE            72.000   
1695521             MCOU    NV    591060701    LABETALOL          6746.000   
1695522             FFSU    MD  63323050830    MEROPENEM           126.000   
1695523             MCOU    GA  59762501601   FLUCONAZOL           216.000   
1695524             FFSU    NC    378180901   LEVOTHYROX          3313.000   
1695525             FFSU    MN    173071820   FLOVENT HF          3309.498   
1695526             FFSU    XX  68084033701   LEVETIRACE           178.000   
1695527             MCOU    FL  31722051290   SIMVASTATI          4347.000   
1695528             FFSU    XX    168033230   FLUTICASON         14355.000   
1695529             FFSU    XX  39822105505   AMPHOTERIC           539.300   
1695530             MCOU    NY  51672404806   CLOTRIMAZO        117675.000   
1695531             FFSU    AK    527134410   LEVOTHYROX          1446.000   
1695532             FFSU    FL    185083101   DEXTROAMPH          1080.000   
1695533             FFSU    OR    777310530       PROZAC           570.000   
1695534             MCOU    WA  51862007001    ESTAZOLAM           460.000   
1695535             FFSU    OR


- Modify `medicaid_data` to uppercase all the column names 

  - If your solution uses an assignment, the righthand side of the assignment (rvalue) can only use `DataFrame` or `Series` methods or properties


In [11]:
medicaid_data.columns = map(str.upper, medicaid_data.columns)
medicaid_data

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
0                   MCOU    PA  55150023930   Dexamethas            33.000   
1                   FFSU    NY     23917710   ALPHAGAN P           570.000   
2                   MCOU    OR  13925050501   Dapsone 10           456.000   
3                   FFSU    MN  51862006401     DIAZEPAM           780.000   
4                   FFSU    MN    781237101   DEXTROAMPH           451.000   
5                   FFSU    VT    603786274   TRIAMCINOL           315.000   
6                   FFSU    NY  24385060271   CHEST CONG          4773.000   
7                   FFSU    MO    603024118       Q-DRYL          9581.000   
8                   FFSU    CT  62756026002   ZONISAMIDE          2530.000   
9                   FFSU    OK  55111065301    Sirolimus           750.000   
10                  FFSU    KY    406012401   HYDROCODON          3564.000   
11                  MCOU    GA  68382000414   Topiramate          8218.000   
12                  FFSU    MA    406012501   HYDROCODON         15732.000   
13                  FFSU    XX    904657561   PHENOBARBI          1576.000   
14                  FFSU    LA  43199003701    LABETALOL          1064.000   
15                  FFSU    MO    472524267   PERMETHRIN          2068.000   
16                  FFSU    ID  67877014601    TEMAZEPAM          2990.000   
17                  FFSU    NV    173087310   INCRUSE EL          1050.000   
18                  MCOU    NY  68645051070   HYDROCHLOR         71982.000   
19                  FFSU    MD    378306677   FENOFIBRAT           724.000   
20                  FFSU    PA    456060010   TEFLARO 60           224.000   
21                  FFSU    XX  18860010501   FAZACLO 15          4206.000   
22                  FFSU    OH  69097015815    MELOXICAM          3426.000   
23                  FFSU    MT    185007460   LOVASTATIN          1421.000   
24                  MCOU    OH    409488799   STERILE WA          4054.601   
25                  MCOU    VA  59762100501   Cabergolin           104.000   
26                  MCOU    FL  29300011316   LAMOTRIGIN           872.000   
27                  FFSU    TN      3633717       DROXIA          6187.000   
28                  FFSU    IN     69046956   Chantix (v           890.000   
29                  FFSU    SC    603497521   OXYBUTYNIN          2747.000   
...                  ...   ...          ...          ...               ...   
1695516             MCOU    MI    168000615   TRIAMCINOL         71290.000   
1695517             MCOU    MI  16714069501   FLUCONAZOL          7340.000   
1695518             MCOU    NE    173069500   ADVAIR DIS         26520.000   
1695519             FFSU    OK  59762444002   Methylpred         17152.000   
1695520             MCOU    FL     54002511   MEFLOQUINE            72.000   
1695521             MCOU    NV    591060701    LABETALOL          6746.000   
1695522             FFSU    MD  63323050830    MEROPENEM           126.000   
1695523             MCOU    GA  59762501601   FLUCONAZOL           216.000   
1695524             FFSU    NC    378180901   LEVOTHYROX          3313.000   
1695525             FFSU    MN    173071820   FLOVENT HF          3309.498   
1695526             FFSU    XX  68084033701   LEVETIRACE           178.000   
1695527             MCOU    FL  31722051290   SIMVASTATI          4347.000   
1695528             FFSU    XX    168033230   FLUTICASON         14355.000   
1695529             FFSU    XX  39822105505   AMPHOTERIC           539.300   
1695530             MCOU    NY  51672404806   CLOTRIMAZO        117675.000   
1695531             FFSU    AK    527134410   LEVOTHYROX          1446.000   
1695532             FFSU    FL    185083101   DEXTROAMPH          1080.000   
1695533             FFSU    OR    777310530       PROZAC           570.000   
1695534             MCOU    WA  51862007001    ESTAZOLAM           460.000   
1695535             FFSU    OR


- Familiarize your self with the data
  - the `NDC` column stands for National Drug Code, a universal product identifier for human drugs in the United States
  
  - The remaining column names are self-explanatory
  
- Explore the number of lines and columns in the data

- Check that your column headers are in uppercase

In [12]:
medicaid_data.shape


(1695546, 10)

* If you explore the  `Location` column for all the entries for which "STATE" value is equal to "HI" you'll notice that all the values are identical

* Are there any states that have more than one value for `Location`. 
  * Hint: think about using a sorted aggregation as part of a split-apply-combine operation to answer this question
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [13]:
unique_location = medicaid_data.groupby(by='STATE', as_index=False).agg({'LOCATION': pd.Series.nunique})
print(unique_location)


   STATE  LOCATION
0     AK         1
1     AL         1
2     AR         1
3     AZ         1
4     CA         1
5     CO         1
6     CT         1
7     DC         1
8     DE         1
9     FL         1
10    GA         1
11    HI         1
12    IA         1
13    ID         1
14    IL         1
15    IN         1
16    KS         1
17    KY         1
18    LA         1
19    MA         1
20    MD         1
21    ME         1
22    MI         1
23    MN         1
24    MO         1
25    MS         1
26    MT         1
27    NC         1
28    ND         1
29    NE         1
30    NH         1
31    NJ         1
32    NM         1
33    NV         1
34    NY         1
35    OH         1
36    OK         1
37    OR         1
38    PA         1
39    RI         1
40    SC         1
41    SD         1
42    TN         1
43    TX         1
44    UT         1
45    VA         1
46    VT         1
47    WA         1
48    WI         1
49    WV         1
50    WY         1
51    XX    



* To compare medication prescriptions across states in a fair and balanced way, we need the number of Medicaid beneficiaries in each state. The following example illustrates the importance of normalizing the values `UNITS REIMBURSED` for each medication in each state by the number of Medicaid enrollees in each state.
  
* The `medicaid_data` DataFrame shows that for the drug with NDC `61958180101` (the drug name is HARVONI and it's used to treat Hepatitis C) there were 11,886  units sold in KY, versus 40,142 in CA -- that's almost 4 times more units sold in CA compared to KY. However, there are 1,284,193 Medicaid enrollees in KY, versus 13,096,861 in California. If we normalize the number of units sold in KY, versus CA, we find that the normalized there were close to 3 times more HARVONI prescription in KY  than in CA. This is ___perhaps___ justified by the fact the KY has one of the highest rates of reported cases of Hepatitis C in the US (2.7% in KY versus 0.2% in CA).
  
https://www.cdc.gov/hepatitis/statistics/2015surveillance/pdfs/2015hepsurveillancerpt.pdf

* The number of enrollees per state was obtained here:

    https://www.medicaid.gov/medicaid/managed-care/enrollment/index.html
    
    
* A parsed/processed version (medicaid_enrollment.tsv) can be in data director of the assignment folder. Use `pandas` to load the medicaid_enrollmen file into DataFrame named `medicaid_enrollment`

In [14]:
medicaid_enrollment = pd.read_csv("data/medicaid_enrollment.tsv", sep='\t', header=0)
medicaid_enrollment

STATE Total Medicaid Enrollees
0                    Alabama                1,050,989
1                     Alaska                  164,783
2             American Samoa                      NaN
3                    Arizona                1,740,520
4                   Arkansas                  762,166
5                 California               13,096,861
6                   Colorado                1,264,600
7                Connecticut                  746,119
8                   Delaware                  227,909
9       District of Columbia                  271,428
10                   Florida                3,808,334
11                   Georgia                1,990,810
12                      Guam                      n/a
13                    Hawaii                  340,513
14                     Idaho                  283,355
15                  Illinois                3,269,999
16                   Indiana                1,295,358
17                      Iowa                  618,505
18                    Kansas                  403,844
19                  Kentucky                1,284,193
20                 Louisiana                1,402,212
21                     Maine                  288,324
22                  Maryland                1,271,445
23             Massachusetts                1,829,618
24                  Michigan                3,947,031
25                 Minnesota                1,052,521
26               Mississippi                  740,937
27                  Missouri                  944,257
28                   Montana                  139,950
29                  Nebraska                  239,463
30                    Nevada                  588,304
31             New Hampshire                  186,399
32                New Jersey                1,705,594
33                New Mexico                  826,155
34                  New York                6,281,038
35            North Carolina                1,965,805
36              North Dakota                   86,250
37  Northern Mariana Islands                      NaN
38                      Ohio                3,060,446
39                  Oklahoma                  829,561
40                    Oregon                1,123,913
41              Pennsylvania                2,569,232
42               Puerto Rico                1,458,819
43              Rhode Island                  308,521
44            South Carolina                1,233,430
45              South Dakota                  124,497
46                 Tennessee                1,562,745
47                     Texas                4,273,982
48                      Utah                  293,867
49                   Vermont                  206,469
50            Virgin Islands                      NaN
51                  Virginia                1,092,225
52                Washington                1,771,679
53             West Virginia                  545,748
54                 Wisconsin                1,209,714
55                   Wyoming                   66,532

* Modify `medicaid_enrollment` to uppercase all the column names 

  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself


In [15]:
medicaid_enrollment.columns = map(str.upper, medicaid_enrollment.columns)
medicaid_enrollment

STATE TOTAL MEDICAID ENROLLEES
0                    Alabama                1,050,989
1                     Alaska                  164,783
2             American Samoa                      NaN
3                    Arizona                1,740,520
4                   Arkansas                  762,166
5                 California               13,096,861
6                   Colorado                1,264,600
7                Connecticut                  746,119
8                   Delaware                  227,909
9       District of Columbia                  271,428
10                   Florida                3,808,334
11                   Georgia                1,990,810
12                      Guam                      n/a
13                    Hawaii                  340,513
14                     Idaho                  283,355
15                  Illinois                3,269,999
16                   Indiana                1,295,358
17                      Iowa                  618,505
18                    Kansas                  403,844
19                  Kentucky                1,284,193
20                 Louisiana                1,402,212
21                     Maine                  288,324
22                  Maryland                1,271,445
23             Massachusetts                1,829,618
24                  Michigan                3,947,031
25                 Minnesota                1,052,521
26               Mississippi                  740,937
27                  Missouri                  944,257
28                   Montana                  139,950
29                  Nebraska                  239,463
30                    Nevada                  588,304
31             New Hampshire                  186,399
32                New Jersey                1,705,594
33                New Mexico                  826,155
34                  New York                6,281,038
35            North Carolina                1,965,805
36              North Dakota                   86,250
37  Northern Mariana Islands                      NaN
38                      Ohio                3,060,446
39                  Oklahoma                  829,561
40                    Oregon                1,123,913
41              Pennsylvania                2,569,232
42               Puerto Rico                1,458,819
43              Rhode Island                  308,521
44            South Carolina                1,233,430
45              South Dakota                  124,497
46                 Tennessee                1,562,745
47                     Texas                4,273,982
48                      Utah                  293,867
49                   Vermont                  206,469
50            Virgin Islands                      NaN
51                  Virginia                1,092,225
52                Washington                1,771,679
53             West Virginia                  545,748
54                 Wisconsin                1,209,714
55                   Wyoming                   66,532

* Note that some states/territories have missing values. Remove the missing values and save the resulting `DataFrame` as a new variable named `medicaid_enrollment_cleaned`

* Pay attention to how 'n/a' is given here!
* After cleaning, do you still have the Guam entry? If so, reconsider what missing values means in this context

In [16]:
medicaid_enrollment_cleaned = medicaid_enrollment.dropna()
medicaid_enrollment_cleaned = medicaid_enrollment_cleaned[~medicaid_enrollment_cleaned["TOTAL MEDICAID ENROLLEES"].str.contains("n/a")]
medicaid_enrollment_cleaned

STATE TOTAL MEDICAID ENROLLEES
0                Alabama                1,050,989
1                 Alaska                  164,783
3                Arizona                1,740,520
4               Arkansas                  762,166
5             California               13,096,861
6               Colorado                1,264,600
7            Connecticut                  746,119
8               Delaware                  227,909
9   District of Columbia                  271,428
10               Florida                3,808,334
11               Georgia                1,990,810
13                Hawaii                  340,513
14                 Idaho                  283,355
15              Illinois                3,269,999
16               Indiana                1,295,358
17                  Iowa                  618,505
18                Kansas                  403,844
19              Kentucky                1,284,193
20             Louisiana                1,402,212
21                 Maine                  288,324
22              Maryland                1,271,445
23         Massachusetts                1,829,618
24              Michigan                3,947,031
25             Minnesota                1,052,521
26           Mississippi                  740,937
27              Missouri                  944,257
28               Montana                  139,950
29              Nebraska                  239,463
30                Nevada                  588,304
31         New Hampshire                  186,399
32            New Jersey                1,705,594
33            New Mexico                  826,155
34              New York                6,281,038
35        North Carolina                1,965,805
36          North Dakota                   86,250
38                  Ohio                3,060,446
39              Oklahoma                  829,561
40                Oregon                1,123,913
41          Pennsylvania                2,569,232
42           Puerto Rico                1,458,819
43          Rhode Island                  308,521
44        South Carolina                1,233,430
45          South Dakota                  124,497
46             Tennessee                1,562,745
47                 Texas                4,273,982
48                  Utah                  293,867
49               Vermont                  206,469
51              Virginia                1,092,225
52            Washington                1,771,679
53         West Virginia                  545,748
54             Wisconsin                1,209,714
55               Wyoming                   66,532

### Converting `TOTAL MEDICAID ENROLLEE` Data Type

* Given that data on `TOTAL MEDICAID ENROLLEE` column contains commas on file (ex. 3,269,999 instead of 3269999), `pandas` has erroneously set the data type for that column as a string. We need to convert the column from string to `int` since we will be using it in an arithmetic expression during normalization

 

* Inspect the `dtype` property of "TOTAL MEDICAID ENROLLEES" column, and  make sure that the data type is `int`


In [17]:
medicaid_enrollment_cleaned["TOTAL MEDICAID ENROLLEES"] = medicaid_enrollment_cleaned["TOTAL MEDICAID ENROLLEES"].str.replace(',', '').astype(int)
medicaid_enrollment_cleaned

STATE  TOTAL MEDICAID ENROLLEES
0                Alabama                   1050989
1                 Alaska                    164783
3                Arizona                   1740520
4               Arkansas                    762166
5             California                  13096861
6               Colorado                   1264600
7            Connecticut                    746119
8               Delaware                    227909
9   District of Columbia                    271428
10               Florida                   3808334
11               Georgia                   1990810
13                Hawaii                    340513
14                 Idaho                    283355
15              Illinois                   3269999
16               Indiana                   1295358
17                  Iowa                    618505
18                Kansas                    403844
19              Kentucky                   1284193
20             Louisiana                   1402212
21                 Maine                    288324
22              Maryland                   1271445
23         Massachusetts                   1829618
24              Michigan                   3947031
25             Minnesota                   1052521
26           Mississippi                    740937
27              Missouri                    944257
28               Montana                    139950
29              Nebraska                    239463
30                Nevada                    588304
31         New Hampshire                    186399
32            New Jersey                   1705594
33            New Mexico                    826155
34              New York                   6281038
35        North Carolina                   1965805
36          North Dakota                     86250
38                  Ohio                   3060446
39              Oklahoma                    829561
40                Oregon                   1123913
41          Pennsylvania                   2569232
42           Puerto Rico                   1458819
43          Rhode Island                    308521
44        South Carolina                   1233430
45          South Dakota                    124497
46             Tennessee                   1562745
47                 Texas                   4273982
48                  Utah                    293867
49               Vermont                    206469
51              Virginia                   1092225
52            Washington                   1771679
53         West Virginia                    545748
54             Wisconsin                   1209714
55               Wyoming                     66532

### Associating `medicaid_data` and `medicaid_enrollment_cleaned`

- We can use the shared State information across both tables to associate both tables (SQL JOIN).
- However,  `medicaid_data` contains two-letter state abbreviations, while `medicaid_enrollment_cleaned` contains the complete state name
  - We need to convert (or append) two-letter state abbreviations to `medicaid_enrollment_cleaned`

- Pandas can read HTML and parse the code for tables. We will use that functionality to read in the state abbreviations from a Wikipedia page.
  - A brief description of what the code does is included in the comments

In [ ]:
# reads all the tables at the given url
# Header=0 instructs pandas to use line 0 as the header (columns)
tables = pd.read_html('https://www.50states.com/abbreviations.htm', header=0)



# We access the desired table by giving it's index.
# Since the URL contain only one table, then we can access that table using index 0
Codes_abbreviations = tables[0]
Codes_abbreviations.head(5)

In [30]:
import requests

url = "https://www.50states.com/abbreviations.htm"
header = {
  "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36",
  "X-Requested-With": "XMLHttpRequest"
}
r = requests.get(url, headers=header)
Codes_abbreviations = pd.read_html(r.text, header=0)[0]
Codes_abbreviations

US State:  Abbreviation:
0                          Alabama             AL
1                           Alaska             AK
2                          Arizona             AZ
3                         Arkansas             AR
4                       California             CA
5                         Colorado             CO
6                      Connecticut             CT
7                         Delaware             DE
8                          Florida             FL
9                          Georgia             GA
10                          Hawaii             HI
11                           Idaho             ID
12                        Illinois             IL
13                         Indiana             IN
14                            Iowa             IA
15                          Kansas             KS
16                        Kentucky             KY
17                       Louisiana             LA
18                           Maine             ME
19                        Maryland             MD
20                   Massachusetts             MA
21                        Michigan             MI
22                       Minnesota             MN
23                     Mississippi             MS
24                        Missouri             MO
25                         Montana             MT
26                        Nebraska             NE
27                          Nevada             NV
28                   New Hampshire             NH
29                      New Jersey             NJ
..                             ...            ...
37                    Pennsylvania             PA
38                    Rhode Island             RI
39                  South Carolina             SC
40                    South Dakota             SD
41                       Tennessee             TN
42                           Texas             TX
43                            Utah             UT
44                         Vermont             VT
45                        Virginia             VA
46                      Washington             WA
47                   West Virginia             WV
48                       Wisconsin             WI
49                         Wyoming             WY
50         Commonwealth/Territory:  Abbreviation:
51                  American Samoa             AS
52            District of Columbia             DC
53  Federated States of Micronesia             FM
54                            Guam             GU
55                Marshall Islands             MH
56        Northern Mariana Islands             MP
57                           Palau             PW
58                     Puerto Rico             PR
59                  Virgin Islands             VI
60               Military "State":  Abbreviation:
61             Armed Forces Africa             AE
62           Armed Forces Americas             AA
63             Armed Forces Canada             AE
64             Armed Forces Europe             AE
65        Armed Forces Middle East             AE
66            Armed Forces Pacific             AP

[67 rows x 2 columns]

* Change the the `DataFrame`'s headers from ['US State:', 'Abbreviation:'] to ['US STATE', 'ABBREVIATION']

  * You can hard code this operation


In [31]:
Codes_abbreviations.columns = ['US STATE', 'ABBREVIATION']
Codes_abbreviations


US STATE   ABBREVIATION
0                          Alabama             AL
1                           Alaska             AK
2                          Arizona             AZ
3                         Arkansas             AR
4                       California             CA
5                         Colorado             CO
6                      Connecticut             CT
7                         Delaware             DE
8                          Florida             FL
9                          Georgia             GA
10                          Hawaii             HI
11                           Idaho             ID
12                        Illinois             IL
13                         Indiana             IN
14                            Iowa             IA
15                          Kansas             KS
16                        Kentucky             KY
17                       Louisiana             LA
18                           Maine             ME
19                        Maryland             MD
20                   Massachusetts             MA
21                        Michigan             MI
22                       Minnesota             MN
23                     Mississippi             MS
24                        Missouri             MO
25                         Montana             MT
26                        Nebraska             NE
27                          Nevada             NV
28                   New Hampshire             NH
29                      New Jersey             NJ
..                             ...            ...
37                    Pennsylvania             PA
38                    Rhode Island             RI
39                  South Carolina             SC
40                    South Dakota             SD
41                       Tennessee             TN
42                           Texas             TX
43                            Utah             UT
44                         Vermont             VT
45                        Virginia             VA
46                      Washington             WA
47                   West Virginia             WV
48                       Wisconsin             WI
49                         Wyoming             WY
50         Commonwealth/Territory:  Abbreviation:
51                  American Samoa             AS
52            District of Columbia             DC
53  Federated States of Micronesia             FM
54                            Guam             GU
55                Marshall Islands             MH
56        Northern Mariana Islands             MP
57                           Palau             PW
58                     Puerto Rico             PR
59                  Virgin Islands             VI
60               Military "State":  Abbreviation:
61             Armed Forces Africa             AE
62           Armed Forces Americas             AA
63             Armed Forces Canada             AE
64             Armed Forces Europe             AE
65        Armed Forces Middle East             AE
66            Armed Forces Pacific             AP

[67 rows x 2 columns]

* Combine the tables `medicaid_enrollment_cleaned` and `Codes_abbreviations` such that the resulting `DataFrame` contains all the columns in `medicaid_enrollment_cleaned` and only `ABBREVIATION` from `Codes_abbreviations` 
* Save the results to variable named `medicaid_enrollment_cleaned_with_zip`

- `medicaid_enrollment_cleaned_with_zip` should look like the following ( '...' represents remaining data that is not shown):


```
   STATE    Total Medicaid Enrollees    ABBREVIATION
0    Alabama    1,050,989    AL
1    Alaska    164,783    AK
2    Arizona    1,740,520    AZ
3    Arkansas    762,166    AR
4    California    13,096,861    CA
...
```

* We did not cover joins in class -- you find a plethora of examples on how to do this online. See for instance:

`https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html`

* If you cannot get it to work, contact the `TA` for the solution. You will not be penalized if you don't answer this question. 

In [32]:
#medicaid_enrollment_cleaned_with_zip = pd.join([medicaid_enrollment_cleaned, Codes_abbreviations["ABBREVIATION"]], axis=1, sort=False)
#medicaid_enrollment_cleaned_with_zip.dropna()
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned.join(Codes_abbreviations.set_index('US STATE'), on='STATE')
medicaid_enrollment_cleaned_with_zip

STATE  TOTAL MEDICAID ENROLLEES ABBREVIATION
0                Alabama                   1050989           AL
1                 Alaska                    164783           AK
3                Arizona                   1740520           AZ
4               Arkansas                    762166           AR
5             California                  13096861           CA
6               Colorado                   1264600           CO
7            Connecticut                    746119           CT
8               Delaware                    227909           DE
9   District of Columbia                    271428           DC
10               Florida                   3808334           FL
11               Georgia                   1990810           GA
13                Hawaii                    340513           HI
14                 Idaho                    283355           ID
15              Illinois                   3269999           IL
16               Indiana                   1295358           IN
17                  Iowa                    618505           IA
18                Kansas                    403844           KS
19              Kentucky                   1284193           KY
20             Louisiana                   1402212           LA
21                 Maine                    288324           ME
22              Maryland                   1271445           MD
23         Massachusetts                   1829618           MA
24              Michigan                   3947031           MI
25             Minnesota                   1052521           MN
26           Mississippi                    740937           MS
27              Missouri                    944257           MO
28               Montana                    139950           MT
29              Nebraska                    239463           NE
30                Nevada                    588304           NV
31         New Hampshire                    186399           NH
32            New Jersey                   1705594           NJ
33            New Mexico                    826155           NM
34              New York                   6281038           NY
35        North Carolina                   1965805           NC
36          North Dakota                     86250           ND
38                  Ohio                   3060446           OH
39              Oklahoma                    829561           OK
40                Oregon                   1123913           OR
41          Pennsylvania                   2569232           PA
42           Puerto Rico                   1458819           PR
43          Rhode Island                    308521           RI
44        South Carolina                   1233430           SC
45          South Dakota                    124497           SD
46             Tennessee                   1562745           TN
47                 Texas                   4273982           TX
48                  Utah                    293867           UT
49               Vermont                    206469           VT
51              Virginia                   1092225           VA
52            Washington                   1771679           WA
53         West Virginia                    545748           WV
54             Wisconsin                   1209714           WI
55               Wyoming                     66532           WY

* We have no further use for the column STATE in  `medicaid_enrollment_cleaned_with_zip`
  * Remove the column make sure your data in `medicaid_enrollment_cleaned_with_zip` looks like the following  ( `...` represents remaining data that is not shown):

```
    Total Medicaid Enrollees    ABBREVIATION
0   1,050,989                  AL
1   164,783                    AK
2   1,740,520                  AZ
3   762,166                    AR
4   13,096,861                 CA
....
```

In [33]:
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.drop('STATE', axis=1)
medicaid_enrollment_cleaned_with_zip

TOTAL MEDICAID ENROLLEES ABBREVIATION
0                    1050989           AL
1                     164783           AK
3                    1740520           AZ
4                     762166           AR
5                   13096861           CA
6                    1264600           CO
7                     746119           CT
8                     227909           DE
9                     271428           DC
10                   3808334           FL
11                   1990810           GA
13                    340513           HI
14                    283355           ID
15                   3269999           IL
16                   1295358           IN
17                    618505           IA
18                    403844           KS
19                   1284193           KY
20                   1402212           LA
21                    288324           ME
22                   1271445           MD
23                   1829618           MA
24                   3947031           MI
25                   1052521           MN
26                    740937           MS
27                    944257           MO
28                    139950           MT
29                    239463           NE
30                    588304           NV
31                    186399           NH
32                   1705594           NJ
33                    826155           NM
34                   6281038           NY
35                   1965805           NC
36                     86250           ND
38                   3060446           OH
39                    829561           OK
40                   1123913           OR
41                   2569232           PA
42                   1458819           PR
43                    308521           RI
44                   1233430           SC
45                    124497           SD
46                   1562745           TN
47                   4273982           TX
48                    293867           UT
49                    206469           VT
51                   1092225           VA
52                   1771679           WA
53                    545748           WV
54                   1209714           WI
55                     66532           WY

- Use `DataFrame medicaid_enrollment_cleaned_with_zip` to assign the appropriate number of Medicaid enrollees to each entry in the `medicaid_data`

   I.E., instead of the 10 original columns, `medicaid_data` will now have an 11th column representing the `TOTAL MEDICAID ENROLLEES` according to the STATE value in the entry.
   
- Save the resulting DataFrame into a new variable called `medicaid_data_w_enrollments`
- The resulting DataFrame should look like the following (`...` represents remaining data that is not shown):

```
UTILIZATION TYPE    STATE    NDC    PRODUCT NAME    UNITS REIMBURSED    NUMBER OF PRESCRIPTIONS    TOTAL AMOUNT REIMBURSED    MEDICAID AMOUNT REIMBURSED    NON MEDICAID AMOUNT REIMBURSED    LOCATION
0    MCOU    PA    55150023930    Dexamethas    33.0    19.0    234.98    234.98    0.0    (40.5773, -77.264)
1    FFSU    NY    23917710    ALPHAGAN P    570.0    57.0    16006.34    16006.34    0.0    (42.1497, -74.9384)
2    MCOU    OR    13925050501    Dapsone 10    456.0    15.0    1052.42    1052.42    0.0    (44.5672, -122.1269)
...
```

* The order of the columns in the `DataFrame` is not important. This answer uses the same approach as the one used to `merge` the tables above. 

In [34]:
medicaid_data_w_enrollments = pd.merge(medicaid_enrollment_cleaned_with_zip, medicaid_data, how='left', left_on = 'ABBREVIATION', right_on = 'STATE').drop('ABBREVIATION', axis=1)
medicaid_data_w_enrollments

TOTAL MEDICAID ENROLLEES UTILIZATION TYPE STATE           NDC  \
0                         1050989             FFSU    AL  5.976250e+10   
1                         1050989             FFSU    AL  6.586200e+10   
2                         1050989             FFSU    AL  4.733507e+10   
3                         1050989             FFSU    AL  3.982204e+10   
4                         1050989             FFSU    AL  3.780473e+08   
5                         1050989             FFSU    AL  6.838201e+10   
6                         1050989             FFSU    AL  9.341507e+07   
7                         1050989             FFSU    AL  4.280600e+10   
8                         1050989             FFSU    AL  5.038309e+10   
9                         1050989             FFSU    AL  6.437005e+10   
10                        1050989             FFSU    AL  5.511107e+10   
11                        1050989             FFSU    AL  5.970024e+08   
12                        1050989             FFSU    AL  9.343590e+07   
13                        1050989             FFSU    AL  3.780782e+08   
14                        1050989             FFSU    AL  7.130638e+08   
15                        1050989             FFSU    AL  6.800102e+10   
16                        1050989             FFSU    AL  1.054200e+10   
17                        1050989             FFSU    AL  9.303500e+07   
18                        1050989             FFSU    AL  7.815915e+08   
19                        1050989             FFSU    AL  4.285800e+10   
20                        1050989             FFSU    AL  6.217502e+10   
21                        1050989             FFSU    AL  5.254410e+10   
22                        1050989             FFSU    AL  5.152578e+10   
23                        1050989             FFSU    AL  9.309000e+07   
24                        1050989             FFSU    AL  6.410121e+08   
25                        1050989             FFSU    AL  5.443013e+07   
26                        1050989             FFSU    AL  5.271352e+08   
27                        1050989             FFSU    AL  3.783506e+08   
28                        1050989             FFSU    AL  9.046215e+08   
29                        1050989             FFSU    AL  2.283083e+08   
...                           ...              ...   ...           ...   
1542264                     66532             FFSU    WY  3.781176e+08   
1542265                     66532             FFSU    WY  5.045806e+10   
1542266                     66532             FFSU    WY  6.787701e+10   
1542267                     66532             FFSU    WY  2.323830e+06   
1542268                     66532             FFSU    WY  5.550974e+08   
1542269                     66532             FFSU    WY  6.516205e+10   
1542270                     66532             FFSU    WY  9.303140e+07   
1542271                     66532             FFSU    WY  1.671404e+10   
1542272                     66532             FFSU    WY  1.730695e+08   
1542273                     66532             FFSU    WY  6.476402e+10   
1542274                     66532             FFSU    WY  6.516201e+10   
1542275                     66532             FFSU    WY  4.596305e+10   
1542276                     66532             FFSU    WY  7.437990e+07   
1542277                     66532             FFSU    WY  6.909709e+10   
1542278                     66532             FFSU    WY  3.785310e+08   
1542279                     66532             FFSU    WY  1.671405e+10   
1542280                     66532             FFSU    WY  6.909702e+10   
1542281                     66532             FFSU    WY  2.282028e+08   
1542282                     66532             FFSU    WY  5.254402e+10   
1542283                     66532             FFSU    WY  5.226800e+10   
1542284                     66532             FFSU    WY  5.723702e+10   
1542285                     66532             FFSU    WY  1.672902e+10   
1542286      

- Remove any lines where "STATE" or "PRODUCT NAME" are missing from  `medicaid_data_w_enrollments`

In [35]:
medicaid_data_w_enrollments = medicaid_data_w_enrollments.dropna(subset=['STATE', 'PRODUCT NAME'])
medicaid_data_w_enrollments[medicaid_data_w_enrollments['NDC'].isnull()]
medicaid_data_w_enrollments



TOTAL MEDICAID ENROLLEES UTILIZATION TYPE STATE           NDC  \
0                         1050989             FFSU    AL  5.976250e+10   
1                         1050989             FFSU    AL  6.586200e+10   
2                         1050989             FFSU    AL  4.733507e+10   
3                         1050989             FFSU    AL  3.982204e+10   
4                         1050989             FFSU    AL  3.780473e+08   
5                         1050989             FFSU    AL  6.838201e+10   
6                         1050989             FFSU    AL  9.341507e+07   
7                         1050989             FFSU    AL  4.280600e+10   
8                         1050989             FFSU    AL  5.038309e+10   
9                         1050989             FFSU    AL  6.437005e+10   
10                        1050989             FFSU    AL  5.511107e+10   
11                        1050989             FFSU    AL  5.970024e+08   
12                        1050989             FFSU    AL  9.343590e+07   
13                        1050989             FFSU    AL  3.780782e+08   
14                        1050989             FFSU    AL  7.130638e+08   
15                        1050989             FFSU    AL  6.800102e+10   
16                        1050989             FFSU    AL  1.054200e+10   
17                        1050989             FFSU    AL  9.303500e+07   
18                        1050989             FFSU    AL  7.815915e+08   
19                        1050989             FFSU    AL  4.285800e+10   
20                        1050989             FFSU    AL  6.217502e+10   
21                        1050989             FFSU    AL  5.254410e+10   
22                        1050989             FFSU    AL  5.152578e+10   
23                        1050989             FFSU    AL  9.309000e+07   
24                        1050989             FFSU    AL  6.410121e+08   
25                        1050989             FFSU    AL  5.443013e+07   
26                        1050989             FFSU    AL  5.271352e+08   
27                        1050989             FFSU    AL  3.783506e+08   
28                        1050989             FFSU    AL  9.046215e+08   
29                        1050989             FFSU    AL  2.283083e+08   
...                           ...              ...   ...           ...   
1542264                     66532             FFSU    WY  3.781176e+08   
1542265                     66532             FFSU    WY  5.045806e+10   
1542266                     66532             FFSU    WY  6.787701e+10   
1542267                     66532             FFSU    WY  2.323830e+06   
1542268                     66532             FFSU    WY  5.550974e+08   
1542269                     66532             FFSU    WY  6.516205e+10   
1542270                     66532             FFSU    WY  9.303140e+07   
1542271                     66532             FFSU    WY  1.671404e+10   
1542272                     66532             FFSU    WY  1.730695e+08   
1542273                     66532             FFSU    WY  6.476402e+10   
1542274                     66532             FFSU    WY  6.516201e+10   
1542275                     66532             FFSU    WY  4.596305e+10   
1542276                     66532             FFSU    WY  7.437990e+07   
1542277                     66532             FFSU    WY  6.909709e+10   
1542278                     66532             FFSU    WY  3.785310e+08   
1542279                     66532             FFSU    WY  1.671405e+10   
1542280                     66532             FFSU    WY  6.909702e+10   
1542281                     66532             FFSU    WY  2.282028e+08   
1542282                     66532             FFSU    WY  5.254402e+10   
1542283                     66532             FFSU    WY  5.226800e+10   
1542284                     66532             FFSU    WY  5.723702e+10   
1542285                     66532             FFSU    WY  1.672902e+10   
1542286      

- Use ["STATE", "PRODUCT NAME"] as hierarchical index for `medicaid_data_w_enrollments`. Recall that a hierarchical index is simply an index with multiple levels of indexing (multiple columns)
  * Hint: the function to set an index on a `DataFrame` can take a single column name or a list of column names. The list here is  ["STATE", "NDC"]
- Call the new data `medicaid_data_w_enrollments_hierarch`
- Inspect your data to make sure the new index has now two levels (STATE and NDC)

In [44]:
medicaid_data_w_enrollments_hierarch = medicaid_data_w_enrollments.set_index(['STATE', 'NDC'])

medicaid_data_w_enrollments_hierarch


TOTAL MEDICAID ENROLLEES UTILIZATION TYPE PRODUCT NAME  \
STATE NDC                                                                    
AL    5.976250e+10                   1050989             FFSU   CLINDAMYCI   
      6.586200e+10                   1050989             FFSU   SERTRALINE   
      4.733507e+10                   1050989             FFSU   TOPIRAMATE   
      3.982204e+10                   1050989             FFSU   Lincomycin   
      3.780473e+08                   1050989             FFSU   DIVALPROEX   
      6.838201e+10                   1050989             FFSU   Pravastati   
      9.341507e+07                   1050989             FFSU   AMOXICILLI   
      4.280600e+10                   1050989             FFSU   Guanfacine   
      5.038309e+10                   1050989             FFSU    Lidocaine   
      6.437005e+10                   1050989             FFSU    Navelbine   
      5.511107e+10                   1050989             FFSU   Lamotrigin   
      5.970024e+08                   1050989             FFSU    Combivent   
      9.343590e+07                   1050989             FFSU    CLOZAPINE   
      3.780782e+08                   1050989             FFSU   FEXOFENADI   
      7.130638e+08                   1050989             FFSU   Ciclopirox   
      6.800102e+10                   1050989             FFSU   LOVASTATIN   
      1.054200e+10                   1050989             FFSU    DIALYVITE   
      9.303500e+07                   1050989             FFSU   ACETAMINOP   
      7.815915e+08                   1050989             FFSU   CHLORPROMA   
      4.285800e+10                   1050989             FFSU    Oxycodone   
      6.217502e+10                   1050989             FFSU   Methylphen   
      5.254410e+10                   1050989             FFSU   ZENCHENT 0   
      5.152578e+10                   1050989             FFSU   Azelastine   
      9.309000e+07                   1050989             FFSU   KETOCONAZO   
      6.410121e+08                   1050989             FFSU   HYDROMORPH   
      5.443013e+07                   1050989             FFSU   FUROSEMIDE   
      5.271352e+08                   1050989             FFSU   LEVOTHYROX   
      3.783506e+08                   1050989             FFSU   RISPERIDON   
      9.046215e+08                   1050989             FFSU   FEXOFENADI   
      2.283083e+08                   1050989             FFSU   ALPRAZOLAM   
...                                      ...              ...          ...   
WY    3.781176e+08                     66532             FFSU    BUSPIRONE   
      5.045806e+10                     66532             FFSU   CONCERTA 3   
      6.787701e+10                     66532             FFSU    Temazepam   
      2.323830e+06                     66532             FFSU    STRATTERA   
      5.550974e+08                     66532             FFSU   DEXTROAMPH   
      6.516205e+10                     66532             FFSU   Benzonatat   
      9.303140e+07                     66532             FFSU    KETOROLAC   
      1.671404e+10                     66532             FFSU    DEBLITANE   
      1.730695e+08                     66532             FFSU   ADVAIR DIS   
      6.476402e+10                     66532             FFSU   DEXILANT 6   
      6.516201e+10                     66532             FFSU   Citalopram   
      4.596305e+10                     66532             FFSU   ONDANSETRO   
      7.437990e+07                     66532             FFSU   HUMIRA - A   
      6.909709e+10                     66532             FFSU   ZONISAMIDE   
      3.785310e+08                     66532             FFSU   ZOLPIDEM T   
      1.671405e+10                     66532             FFSU   CLONAZEPAM   
      6.909702e+10                     66532             FFSU   LAMOTRIGIN   
      2.282028e+08                     66532             FFSU   ALPRAZOLAM   
      5.254402e+10                     665


* Write a single Pandas expression to print all the lines with containing NDC 61958180101 in "PA"

 * Use a single indexing call (bracket notation) using `loc`
 
 * Hint 1: Since your index is hierarchical, `loc` is expecting two values, the first for STATE and the second for NDC


In [145]:
medicaid_data_w_enrollments_hierarch.loc[("PA", 61958180101)]

/home/chris/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: PerformanceWarning: indexing past lexsort depth may impact performance.
  """Entry point for launching an IPython kernel.


TOTAL MEDICAID ENROLLEES UTILIZATION TYPE PRODUCT NAME  \
STATE NDC                                                                    
PA    6.195818e+10                   2569232             FFSU   Harvoni  (   
      6.195818e+10                   2569232             MCOU   Harvoni  (   
      6.195818e+10                   2569232             FFSU   Harvoni  (   
      6.195818e+10                   2569232             MCOU   Harvoni  (   
      6.195818e+10                   2569232             FFSU   Harvoni  (   
      6.195818e+10                   2569232             MCOU   Harvoni  (   

                    UNITS REIMBURSED  NUMBER OF PRESCRIPTIONS  \
STATE NDC                                                       
PA    6.195818e+10             924.0                     33.0   
      6.195818e+10            2604.0                     93.0   
      6.195818e+10            1008.0                     36.0   
      6.195818e+10            1932.0                     69.0   
      6.195818e+10             924.0                     33.0   
      6.195818e+10            3220.0                    115.0   

                    TOTAL AMOUNT REIMBURSED  MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                 
PA    6.195818e+10               1017644.54                   985269.84   
      6.195818e+10               2952891.16                  2860100.54   
      6.195818e+10               1107725.22                  1107725.22   
      6.195818e+10               2178602.30                  2118612.30   
      6.195818e+10               1015383.60                  1015383.60   
      6.195818e+10               3599192.35                  3482188.68   

                    NON MEDICAID AMOUNT REIMBURSED            LOCATION  
STATE NDC                                                               
PA    6.195818e+10                        32374.70  (40.5773, -77.264)  
      6.195818e+10                        92790.62  (40.5773, -77.264)  
      6.195818e+10                            0.00  (40.5773, -77.264)  
      6.195818e+10                        59990.00  (40.5773, -77.264)  
      6.195818e+10                            0.00  (40.5773, -77.264)  
      6.195818e+10                       117003.67  (40.5773, -77.264)

#### Computing the normalized `UNITS REIMBURSED` per druc 

* Compute the `UNITS REIMBURSED` for each "NDC" in each state normalized by the number of enrollees.

- For instance in `PA`, the total `UNITS REIMBURSED` for the HARVONI `NDC` (61958180101) is 10,612

```python
total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.loc[("PA", 61958180101), "UNITS REIMBURSED"].sum() 
```

- And the numebr of Medicaid Enrollees in "PA" is 2,569,232


```python
total_enrollees_PA = medicaid_data_w_enrollments_hierarch.loc["PA", "TOTAL MEDICAID ENROLLEES"].unique()[0]
```
- Therefore, the UNITS REIMBURSED per enrollee  for "HARVONI" is  0.00413041718303


```python
print(total_amount_reimbursed/total_enrollees_AK)
```

- Rather than work directly with the ratios, we are going to compute and work with thier logarithm (np.log2) instead.

  - The reason we use logs here is to avoid working small numbers. More on log-transformation in future lectures

- Save the result in a `Series` called `medicaid_reimbursement_per_enrollee`


- Your final result should be a Series that look like the following ( `...` represents data that is not shown here):
  
```
STATE  NDC    
AK     2143380    -9.609109
       2143480   -10.008280
       2322730    -6.109830
       2322830    -4.444321
       2322930    -3.855995
...

AL     2143380    -9.940595
       2143480    -9.805485
       2322730    -5.336260
...

MA     2143380    -7.921997
       2143480    -7.803463
       2144511   -13.472194
       2197590    -7.741402
       2322730    -5.414724
       2322830    -4.592154
       2322930    -4.205626

...


```


In [307]:
#total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.groupby(level=[0, 1]).sum()
#total_amount_reimbursed.iloc[:, 0:2]
#total_enrollees_PA = medicaid_data_w_enrollments_hierarch.groupby(["STATE", "TOTAL MEDICAID ENROLLEES"]).nunique()
#total_enrollees_PA.iloc[:, 0:0]
total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.iloc[:, 3]
total_amount_reimbursed = total_amount_reimbursed.groupby(level=[0, 1]).sum()
print(isinstance(total_amount_reimbursed, pd.Series))
total_enrollees = medicaid_data_w_enrollments_hierarch.iloc[:, 0]
total_enrollees = total_enrollees.groupby(level=[0, 1]).unique()
print(isinstance(total_enrollees, pd.Series))

medicaid_reimbursement_per_enrollee = pd.concat([total_amount_reimbursed, total_enrollees], axis=1)

medicaid_reimbursement_per_enrollee['RESULT'] = medicaid_reimbursement_per_enrollee.iloc[:, 0]/medicaid_reimbursement_per_enrollee.iloc[:, 1]

print(isinstance(medicaid_reimbursement_per_enrollee, pd.DataFrame))
medicaid_reimbursement_per_enrollee

##################################################################################################################

#print(medicaid_data_w_enrollments_hierarch.iloc[0, 0:1])
#total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.loc[("AL", 59762501002), "UNITS REIMBURSED"].sum()
#total_amount_reimbursed
#total_enrollees_PA = medicaid_data_w_enrollments_hierarch.loc["AL", "TOTAL MEDICAID ENROLLEES"].unique()[0]
#total_enrollees_PA

#medicaid_reimbursement_per_enrollee = total_amount_reimbursed/total_enrollees
#medicaid_reimbursement_per_enrollee


True
True
True


UNITS REIMBURSED TOTAL MEDICAID ENROLLEES  \
STATE NDC                                                       
AK    2.143380e+06           211.000                 [164783]   
      2.143480e+06           160.000                 [164783]   
      2.322730e+06          2386.000                 [164783]   
      2.322830e+06          7569.000                 [164783]   
      2.322930e+06         11380.000                 [164783]   
      2.323830e+06          3292.000                 [164783]   
      2.323930e+06          5412.000                 [164783]   
      2.325030e+06          3662.000                 [164783]   
      2.325130e+06          1410.000                 [164783]   
      2.512330e+06          4683.000                 [164783]   
      2.750101e+06            45.000                 [164783]   
      2.750201e+06            36.780                 [164783]   
      2.751001e+06         15960.000                 [164783]   
      2.751659e+06          1359.000                 [164783]   
      2.762301e+06            64.325                 [164783]   
      2.764001e+06            78.000                 [164783]   
      2.771227e+06           534.000                 [164783]   
      2.771559e+06           255.000                 [164783]   
      2.803101e+06           424.508                 [164783]   
      2.821501e+06           880.000                 [164783]   
      2.821517e+06            71.550                 [164783]   
      2.831501e+06           960.000                 [164783]   
      2.840001e+06           340.800                 [164783]   
      2.871501e+06           530.000                 [164783]   
      2.879759e+06           960.000                 [164783]   
      2.879959e+06         24798.000                 [164783]   
      2.880359e+06           243.000                 [164783]   
      2.880559e+06           735.000                 [164783]   
      3.029305e+06          4504.740                 [164783]   
      3.029320e+06           341.500                 [164783]   
...                              ...                      ...   
WY    6.945201e+10           540.000                  [66532]   
      6.945201e+10          2301.000                  [66532]   
      6.945201e+10           986.000                  [66532]   
      6.945201e+10          1660.000                  [66532]   
      6.945202e+10          1483.000                  [66532]   
      6.954301e+10          1125.000                  [66532]   
      6.954301e+10           248.000                  [66532]   
      6.954301e+10         10190.000                  [66532]   
      6.954303e+10         39589.000                  [66532]   
      6.954303e+10          5376.000                  [66532]   
      6.991802e+10         12106.000                  [66532]   
      7.006900e+10            18.000                  [66532]   
      7.006900e+10           138.480                  [66532]   
      7.006901e+10           115.000                  [66532]   
      7.040801e+10         19909.000                  [66532]   
      7.070003e+10           360.000                  [66532]   
      7.583401e+10          2616.000                  [66532]   
      7.585403e+10          1170.000                  [66532]   
      7.604501e+10           276.500                  [66532]   
      7.620402e+10         42714.000                  [66532]   
      7.620402e+10          5310.000                  [66532]   
      7.620402e+10         72715.200                  [66532]   
      7.620406e+10         17460.000                  [66532]   
      7.620406e+10         39816.000                  [66532]   
      7.632915e+10           184.000                  [66532]   
      7.632934e+10            76.000                  [66532]   
      7.638501e+10           939.000                  [66532]   
      7.638501e+10          2546.000                  [66532]   
      7.638501e+10          1065.000                  [

- To facilitate working with the final data, we are going to unstack `medicaid_reimbursement_per_enrollee` into a variable called  `medicaid_norm_ndc`

- Using `medicaid_reimbursement_per_enrollee`, generate a `DataFrame` where: 
  - index should be the two-letter state symbol 
  - the column names should be the NDC codes 

- The `DataFrame`  should be formatted as in the image below
  - Hint, simply unstack the data


<img src="media/unstacked.png" alt="drawing" style="width:900px;"/>


In [ ]:
## WRITE YOUR CODE HERE 


#### Exploring the data (very briefly) 
- What is the drug with the highest log-normalized ratio in Hawaii?


In [ ]:
## WRITE YOUR CODE HERE 


* Investigate the NDC of the product with the highest log-normalized ratio in Hawaii 
  * What is it used for?

* Compare the value of `Units Reimbursed` that product between HI and other states, (take for instance MA, FL, OR and WA)
* Can you think for reasons why this product has the highest log-normalized ratio in Hawaii?

In [ ]:
## WRITE YOUR CODE HERE 

In [ ]:
## WRITE YOUR ANSWER HERE 

* Find and list all unique `NDC`s for which the difference between the largest and second large log-normalized ratio by state is at least 10.

* For instance:
  * The highest log-normalized ratio for `00591289749` (`AZACITIDIN`) is OK where it has a log-normalized ratio  of `-1.025642`.
  * The second highest log-normalized ratio for `00591289749` is in `GA` where is has a log-normalized ration of  `-12.623428`
  


In [ ]:
## WRITE YOUR CODE HERE 


- The Drug `AZACITIDINE` has a very high normalized UNITS REIMBURSED in OK compared to other states.
   - Normalized log value is -1.025642 (or a ratio 0.49119167009735065)
   - Second highest state has a log value of -12.623428 (0.000158478197834722)
- Oklahoma is not a high-incidence state for cancer
- Could the following explain what is happening in Oklahoma?

https://www.centerwatch.com/clinical-trials/listings/92093/acute-myeloid-leukemia-aml-study-asp2215-gilteritinib-by/?&geo_lat=35.4675602&geo_lng=-97.5164276&radius=10